services/st_util/StUartBootloaderCommandHandler.cpp

```diff
@@ -0,0 +1,552 @@
+    {
+        serial.SendData(data, [this, checksum]()
+            {
+                serial.SendData(infra::MakeConstByteRange(checksum), [this]()
+                    {
+                        TryHandleTransmitAction();
+                    });
+            });
+    }
+
+    void StUartBootloaderCommandHandler::SendData(infra::ConstByteRange data)
+    {
+        serial.SendData(data, [this]()
+            {
+                TryHandleTransmitAction();
+            });
+    }
+
+    void StUartBootloaderCommandHandler::DataReceived()
+    {
+        if (!commandActions.empty())
+            commandActions.front()->DataReceived();
+    }
+
+    void StUartBootloaderCommandHandler::OnError(infra::BoundedConstString reason)
+    {
+        timeout.Cancel();
+        serial.ReceiveData([](auto data) {});
+        if (onError)
+            onError(reason);
+    }
+
+    StUartBootloaderCommandHandler::Action::Action(StUartBootloaderCommandHandler& handler, StUartBootloaderCommandHandler::ActionType type)
+        : handler(handler)
+        , type(type)
+    {}
+
+    void StUartBootloaderCommandHandler::Action::SendData()
+    {
+        std::abort();
+    }
+
+    void StUartBootloaderCommandHandler::Action::DataReceived()
+    {
+        std::abort();
+    }
+
+    StUartBootloaderCommandHandler::ActionType StUartBootloaderCommandHandler::Action::GetActionType() const
+    {
+        return type;
+    }
+
+    StUartBootloaderCommandHandler::ReceiveAction::ReceiveAction(StUartBootloaderCommandHandler& handler)
+        : StUartBootloaderCommandHandler::Action(handler, StUartBootloaderCommandHandler::ActionType::receive)
+    {}
+
+    StUartBootloaderCommandHandler::TransmitAction::TransmitAction(StUartBootloaderCommandHandler& handler)
+        : StUartBootloaderCommandHandler::Action(handler, StUartBootloaderCommandHandler::ActionType::transmit)
+    {}
+
+    void StUartBootloaderCommandHandler::ReceiveAckAction::DataReceived()
+    {
+        auto byte = handler.queue.Get();
+
+        if (byte == ack)
+            handler.TryHandleNextAction();
+        else
+            handler.OnError("NACK received");
+    }
+
+    StUartBootloaderCommandHandler::ReceiveBufferAction::ReceiveBufferAction(StUartBootloaderCommandHandler& handler, infra::ByteRange& data)
+        : StUartBootloaderCommandHandler::ReceiveAction(handler)
+        , data(data)
+    {}
+
+    StUartBootloaderCommandHandler::ReceiveBufferAction::ReceiveBufferAction(StUartBootloaderCommandHandler& handler, infra::ByteRange& data, const std::size_t nBytesTotal)
+        : StUartBootloaderCommandHandler::ReceiveAction(handler)
+        , data(data)
+    {
+        this->nBytesTotal.Emplace(nBytesTotal);
+    }
+
+    void StUartBootloaderCommandHandler::ReceiveBufferAction::DataReceived()
+    {
+        if (handler.queue.Empty())
+            return;
+
+        infra::ByteInputStream stream(handler.queue.ContiguousRange());
```

Comment:
ContiguousRange will only return the range that is contiguous in memory, so it might just return 1 byte. Instead of using ByteInputStream on ContiguousRange, use:
infra::QueueForOneReaderOneIrqWriter<uint8_t>::StreamReader reader(handler.queue);
infra::DataInputStream::WithErrorPolicy stream(reader);